### PR TITLE
Made Idx a newtype over Int

### PR DESCRIPTION
**Pull request by dpvanbalen**
**Description**
See title

**Motivation and context**
Improves compilation and simplification performance.

**How has this been tested?**
Tests pass, and Tom's huge AD program went from 5.6s to 1.6s (Accelerate compile + runtime).

I couldn't test the 'internal-checks' version, because it errors in Analysis/Match (seems unrelatedly broken by some other change, missing a Typeable constraint), but it should be identical to the 'old' version.

**Types of changes**
What types of changes does your code introduce? Put an `x` in all the boxes that apply:

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

**Checklist**
Go over all the following points, and put an `x` in all the boxes that apply. If you're unsure about any of these, don't hesitate to ask. We're here to help!

- [x] My code follows the code style of this project
- [x] My change requires a change to the documentation
- [x] I have updated the documentation accordingly
- [ ] I have added tests to cover my changes
- [x] All new and existing tests passed

### CODE REVIEW

File: src/Data/Array/Accelerate/AST/Idx.hs [hunk 1]
```diff
@@ -38,6 +46,54 @@ rnfIdx :: Idx env t -> ()
...
+-- | Despite the 'complete' pragma above, GHC can't infer that there is no
+-- pattern possible if the environment is empty. This can be used instead.
+pattern NoIdxPossible :: forall env t a. (env ~ ()) => () => a -> Idx env t
+pattern NoIdxPossible a <- (\case{} -> a)
+
+data PairIdx p a where
+  PairIdxLeft  :: PairIdx (a, b) a
+  PairIdxRight :: PairIdx (a, b) b
+
 liftIdx :: Idx env t -> Q (TExp (Idx env t))
```
Comment (dpvanbalen) [inline src/Data/Array/Accelerate/AST/Idx.hs:97]: I'm not sure enough what this is used for -- should this also change?

File: src/Data/Array/Accelerate/AST/Idx.hs [hunk 2]
```diff
@@ -17,19 +25,19 @@ module Data.Array.Accelerate.AST.Idx
   where
 
 import Language.Haskell.TH
+import Data.Type.Equality ((:~:)(Refl))
+import Unsafe.Coerce (unsafeCoerce)
```
Comment (tomsmeding) [inline src/Data/Array/Accelerate/AST/Idx.hs:29]: Do these imports generate "unused import" warnings when in internal-checks mode? If so, maybe put them inside an `#ifndef` in order to not introduce spurious warnings.

File: src/Data/Array/Accelerate/AST/Idx.hs [hunk 3]
```diff
@@ -17,19 +25,19 @@ module Data.Array.Accelerate.AST.Idx
   where
```
Comment (tomsmeding) [inline src/Data/Array/Accelerate/AST/Idx.hs:25]: Maybe explicitly list exports here, and exclude `UnsafeIdxConstructor`? You don't seem to need it anywhere (yay!) and it closes a pretty gaping hole in the type-safety provided by having this in its own module.

File: src/Data/Array/Accelerate/AST/Idx.hs [hunk 4]
```diff
@@ -38,6 +46,54 @@ rnfIdx :: Idx env t -> ()
...
+pattern ZeroIdx :: forall env envt t. () => (envt ~ (env, t)) => Idx envt t
+pattern ZeroIdx <- (\x -> (idxToInt x, unsafeCoerce @(envt :~: envt) @(envt :~: (env, t)) Refl) -> (0, Refl))
+  where
+    ZeroIdx = UnsafeIdxConstructor 0
+
+pattern SuccIdx :: () => (envs ~ (env, s)) => Idx env t -> Idx envs t
+pattern SuccIdx idx <- (unSucc -> Just (idx, Refl))
+  where
+    SuccIdx (UnsafeIdxConstructor i) = UnsafeIdxConstructor (i+1)
+unSucc :: Idx envs t -> Maybe (Idx env t, envs :~: (env, s))
```
Comment (tomsmeding) [inline src/Data/Array/Accelerate/AST/Idx.hs:73]: Newline before the definition of `unSucc`?

File: src/Data/Array/Accelerate/AST/Idx.hs [hunk 5]
```diff
@@ -38,6 +46,54 @@ rnfIdx :: Idx env t -> ()
...
+  | i < 1     = Nothing
+  | otherwise = Just (UnsafeIdxConstructor (i-1), unsafeCoerce Refl)
+
+idxToInt :: Idx env t -> Int
+idxToInt = unsafeRunIdx
+
+rnfIdx :: Idx env t -> ()
+rnfIdx (UnsafeIdxConstructor i)
+  | i > 0     = ()
+  | otherwise = ()
```
Comment (tomsmeding) [inline src/Data/Array/Accelerate/AST/Idx.hs:84]: Why not:

```haskell
rnfIdx :: Idx env t -> ()
rnfIdx !_ = ()
```

given that the new `Idx` is a newtype anyway :)

File: src/Data/Array/Accelerate/AST/Idx.hs [hunk 6]
```diff
@@ -38,6 +46,54 @@ rnfIdx :: Idx env t -> ()
...
+pattern NoIdxPossible :: forall env t a. (env ~ ()) => () => a -> Idx env t
+pattern NoIdxPossible a <- (\case{} -> a)
+
+data PairIdx p a where
+  PairIdxLeft  :: PairIdx (a, b) a
+  PairIdxRight :: PairIdx (a, b) b
+
 liftIdx :: Idx env t -> Q (TExp (Idx env t))
 liftIdx ZeroIdx      = [|| ZeroIdx ||]
 liftIdx (SuccIdx ix) = [|| SuccIdx $$(liftIdx ix) ||]
```
Comment (tomsmeding) [inline src/Data/Array/Accelerate/AST/Idx.hs:99]:
```haskell
liftIdx :: Idx env t -> Q (TExp (Idx env t))
liftIdx (UnsafeIdxConstructor i) = [|| UnsafeIdxConstructor $$(i) ||]
```

I may be missing something though :)

File: src/Data/Array/Accelerate/AST/Idx.hs (hunk 1 above)
Comment (tomsmeding) [reply src/Data/Array/Accelerate/AST/Idx.hs:97]: The TH stuff is for `runQ` AFAIK

File: src/Data/Array/Accelerate/AST/Idx.hs (hunk 2 above)
Comment (dpvanbalen) [reply src/Data/Array/Accelerate/AST/Idx.hs:29]: They do, but I'm not sure whether some 'unused import' warnings on a rare internal flag are really worse than multiple CPP's that depend on each other? Maybe I'm too allergic to these #ifdef's

File: src/Data/Array/Accelerate/AST/Idx.hs (hunk 3 above)
Comment (dpvanbalen) [reply src/Data/Array/Accelerate/AST/Idx.hs:25]: Yes, I'm glad we don't need it! I'm not sure yet whether I will want to use them in my AST rebuilding after fusion, but I'll add it and we can always export it again later

File: src/Data/Array/Accelerate/AST/Idx.hs (hunk 2 above)
Comment (tomsmeding) [reply src/Data/Array/Accelerate/AST/Idx.hs:29]: @tmcdonell I believe other files do try to make sure that setting flags doesn't cause unused import warnings; your thoughts?

File: src/Data/Array/Accelerate/AST/Idx.hs [hunk 7]
```diff
@@ -1,5 +1,14 @@
...
+{-# LANGUAGE CPP #-}
+{-# LANGUAGE EmptyCase #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE PatternSynonyms #-}
+{-# LANGUAGE ScopedTypeVariables #-}
 {-# LANGUAGE TemplateHaskell #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeOperators #-}
+{-# LANGUAGE ViewPatterns #-}
```
Comment (tmcdonell) [inline src/Data/Array/Accelerate/AST/Idx.hs:11]:
```suggestion
{-# LANGUAGE BangPatterns        #-}
{-# LANGUAGE CPP                 #-}
{-# LANGUAGE EmptyCase           #-}
{-# LANGUAGE GADTs               #-}
{-# LANGUAGE LambdaCase          #-}
{-# LANGUAGE PatternSynonyms     #-}
{-# LANGUAGE ScopedTypeVariables #-}
{-# LANGUAGE TemplateHaskell     #-}
{-# LANGUAGE TypeApplications    #-}
{-# LANGUAGE TypeOperators       #-}
{-# LANGUAGE ViewPatterns        #-}
```

File: src/Data/Array/Accelerate/AST/Idx.hs [hunk 8]
```diff
@@ -13,23 +22,24 @@
 -- Typed de Bruijn indices
 --
 
-module Data.Array.Accelerate.AST.Idx
+module Data.Array.Accelerate.AST.Idx 
+  ( Idx, pattern ZeroIdx, pattern SuccIdx, idxToInt, rnfIdx, liftIdx, pattern NoIdxPossible, PairIdx(..) )
   where
```
Comment (tmcdonell) [inline src/Data/Array/Accelerate/AST/Idx.hs:27]:
```suggestion
module Data.Array.Accelerate.AST.Idx (

  Idx, pattern ZeroIdx, pattern SuccIdx, pattern NoIdxPossible,
  idxToInt,
  rnfIdx, liftIdx,

  PairIdx(..)

) where
```

File: src/Data/Array/Accelerate/AST/Idx.hs [hunk 9]
```diff
@@ -42,3 +52,54 @@ liftIdx :: Idx env t -> Q (TExp (Idx env t))
...
+
+liftIdx :: Idx env t -> Q (TExp (Idx env t))
+liftIdx (UnsafeIdxConstructor i) = [|| UnsafeIdxConstructor i ||]
+
+#endif
+
+-- | Despite the 'complete' pragma above, GHC can't infer that there is no
+-- pattern possible if the environment is empty. This can be used instead.
+pattern NoIdxPossible :: forall env t a. (env ~ ()) => () => a -> Idx env t
+pattern NoIdxPossible a <- (\case{} -> a)
```
Comment (tmcdonell) [inline src/Data/Array/Accelerate/AST/Idx.hs:99]: Changing the name to `VoidIdx` would make a nice symmetry in the naming...

**Review by tomsmeding — APPROVED** [review]

**Comment by tmcdonell** [comment]
Thanks @dpvanbalen and @tomsmeding !

**Comment by tmcdonell** [comment]
From the archives: https://github.com/tmcdonell/accelerate/commit/8897d1bf89b3761bef0328e1eac0285ebd718f95.

**Comment by dpvanbalen** [comment]
> From the archives: [tmcdonell@8897d1b](https://github.com/tmcdonell/accelerate/commit/8897d1bf89b3761bef0328e1eac0285ebd718f95).

Interesting! A hybrid approach which retains more internal safety, but SuccIdx' is still recursive (so it only optimises for idxToInt)

**Pull request merged**